### PR TITLE
[Prototype] Make RazorPages work E2E

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

**Comment by pranavkm** [comment]
cc @rynowak

File: src/Microsoft.AspNetCore.Mvc.Razor.Host/BaseDocumentClassifierPass.cs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using Microsoft.AspNetCore.Razor.Evolution;
+using Microsoft.AspNetCore.Razor.Evolution.Intermediate;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Host
+{
+    public abstract class BaseDocumentClassifierPass : IRazorIRPass
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/BaseDocumentClassifierPass.cs]: We should put this in Razor. Most document classifiers we'd write will do 80% of the stuff the same.

File: src/Microsoft.AspNetCore.Mvc.Razor.Host/RazorPageDocumentClassifier.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Razor.Evolution;
+using Microsoft.AspNetCore.Razor.Evolution.Intermediate;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Host
+{
+    public class RazorPageDocumentClassifier : BaseDocumentClassifierPass
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/RazorPageDocumentClassifier.cs]: This is beautiful. Exactly what I had in mind 😻

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorProject.cs [hunk 3]
```diff
@@ -34,6 +34,27 @@ public override IEnumerable<RazorProjectItem> EnumerateItems(string path)
             return EnumerateFiles(_provider.GetDirectoryContents(path), path, "");
         }
 
+        public RazorProjectItem GetItem(string path)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRazorProject.cs]: Should we provide a base-class for this that handles argument validation?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/MvcRazorPagesMvcBuilderExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            return builder;
+        }
+
+        /// <summary>
+        /// Registers tag helpers as services and replaces the existing <see cref="ITagHelperActivator"/>
+        /// with an <see cref="ServiceBasedTagHelperActivator"/>.
+        /// </summary>
+        /// <param name="builder">The <see cref="IMvcBuilder"/> instance this method extends.</param>
+        /// <returns>The <see cref="IMvcBuilder"/> instance this method extends.</returns>
+        public static IMvcBuilder AddTagHelpersAsServices(this IMvcBuilder builder)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/DependencyInjection/MvcRazorPagesMvcBuilderExtensions.cs]: Why is this here?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            IRazorCompilationService razorCompilationService,
+            RazorProject razorProject)
+        {
+            _razorCompilationService = razorCompilationService;
+            _project = razorProject;
+        }
+
+        public Type Load(PageActionDescriptor actionDescriptor)
+        {
+            // Todo : Add more APIs to RazorProject
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs]: No todos 👎

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/MvcRazorPagesRazorViewEngineOptionsSetup.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        /// <param name="options">The <see cref="MvcViewOptions"/> to configure.</param>
+        public void Configure(RazorViewEngineOptions options)
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            var assemblyLocation = GetType().GetTypeInfo().Assembly.Location;
+            options.AdditionalCompilationReferences.Add(MetadataReference.CreateFromFile(assemblyLocation));
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/MvcRazorPagesRazorViewEngineOptionsSetup.cs]: Why is this needed?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Page.cs [hunk 7]
```diff
@@ -41,13 +49,84 @@ public abstract class Page : IRazorPage
...
+            {
+                var services = ViewContext.HttpContext.RequestServices;
+                var factory = services.GetRequiredService<IUrlHelperFactory>();
+                _urlHelper = factory.GetUrlHelper(ViewContext);
+            }
+
+            return _urlHelper.Content(contentPath);
+        }
+
+        public override void DefineSection(string name, RenderAsyncDelegate section)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Page.cs]: lol

File: src/Microsoft.AspNetCore.Mvc.Razor.Host/PageDirective.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+        private class Visitor : RazorIRNodeWalker
+        {
+            public DirectiveIRNode DirectiveNode { get; private set;  }
+
+            public string RouteTemplate { get; private set; }
+
+            public override void VisitDirective(DirectiveIRNode node)
+            {
+                if (node.Descriptor == DirectiveDescriptor)
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/PageDirective.cs]: Would just compare names.
